### PR TITLE
GAP-2452: Add delete to multiple choice and multiple select options

### PR DESCRIPTION
**Pull request by paul-lawlor-tco**
Added delete button to multiple choice/select questions (as seen when editing a question)

GAP-2452: https://technologyprogramme.atlassian.net/browse/GAP-2452

Summary of the changes and the related issue. List any dependencies that are required for this change:

## Type of change

Please check the relevant options.

- [X] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [X] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

MULTIPLE SELECT (EDIT VIEW):
Screenshot 2024-03-05 at 15 20 19

MULTIPLE CHOICE (EDIT VIEW):
Screenshot 2024-03-05 at 15 21 08

# Checklist:

- [ ] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [ ] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

### CODE REVIEW

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx [hunk 1]
```diff
@@ -60,15 +60,25 @@ export const getServerSideProps: GetServerSideProps = async ({
     req,
     res,
     async (body: any) => {
-      options = body.options;
+      options = Object.keys(body).reduce((array, key) => {
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx:63]: I think there is a lot going in here now and might be more readable to split into functions. 

For example one function for processing the options and another for processing the body.

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx [hunk 2]
```diff
@@ -60,15 +60,25 @@ export const getServerSideProps: GetServerSideProps = async ({
...
+      }, [] as string[]);
 
       if ('add-another-option' in body) {
         options.push('');
 
         return {
           data: 'OPTION_ADDED',
         };
-      } else {
+      } else if ('save-question' in body) {
```
Comment (dylanwrightCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx:81]: Given there are a couple more scenarios here we could use a switch statement to improve readability? what do you think?

Comment (ConorFayleTCO) [reply packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx]: the switch doesn't help in this case as we're checking for keys in an object - it means we need to use `switch(true)` like in the other place we're doing this which i intend to refactor out anyway

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx [hunk 3]
```diff
@@ -56,21 +56,32 @@ export const getServerSideProps: GetServerSideProps = async ({
...
+      }
+      return array;
+    }, [] as string[]);
+    return options;
+  }
+
+  async function handleOptions(body: any, options: string[]) {
+    const { optional, ...restOfQuestionSummary } = questionSummary;
+    // switch expression 'true' for evaluation of boolean case labels
+    switch (true) {
```
Comment (ConorFayleTCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx:77]: i think this should be an if-else, it reduces the amount of indentation with no cost to readability

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.test.tsx [hunk 4]
```diff
@@ -155,6 +155,24 @@ describe('Question Options', () => {
       );
       expect(document.title).toBe('Error: Add a question - Manage a grant');
     });
+
+    it('Should NOT render "Delete" button for each option when there are 2 or less options present', () => {
```
Comment (ConorFayleTCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.test.tsx:159]: also need tests for actually deleting options

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx [hunk 5]
```diff
@@ -56,31 +56,32 @@ export const getServerSideProps: GetServerSideProps = async ({
...
-        }
-        return array;
-      }, [] as string[]);
+      if (key.startsWith('delete_')) {
+        const deleteOptionIndex = Number(key.split('_')[1]);
+        array.splice(deleteOptionIndex, 1);
+      }
+      return array;
+    }, [] as string[]);
+    return options;
```
Comment (ConorFayleTCO) [inline packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx:71]:
```suggestion
  return Object.keys(body).reduce((array, key) => {
      if (key.startsWith('options')) {
        array.push(...body[key]);
      } else if (key.startsWith('delete_')) {
        const deleteOptionIndex = Number(key.split('_')[1]);
        array.splice(deleteOptionIndex, 1);
      }
      return array;
    }, [] as string[]);
```
best practice to not directly modify parameters passed to a function unless there's a specific reason, plus using an else-if here reduces complexity and processing slightly

File: packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx (hunk 2 above)
Comment (dylanwrightCO) [reply packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx]: From what i've read using `switch(true)` is pretty widely accepted and not considered bad practice. Would say the readability of this is pretty subjective too as neither is right or wrong. 

If it was any more than 3 cases i'd say go for switch over if else but it's up to you @paul-lawlor-tco

Comment (paul-lawlor-tco) [reply packages/admin/src/pages/build-application/[applicationId]/[sectionId]/question-options.page.tsx]: appreciate that the switch(true) is a bit strange, although having the labels explicitly written out for each case makes it easier to read what's going on imo (regardless of the extra indentation)

will keep it as the switch for now and if you're refactoring then feel free! 

Thank you both for the input on this :)

**Review by dylanwrightCO — APPROVED** [review]

**Pull request merged**